contracts/test/integration/MultiSubnet.t.sol

```diff
@@ -677,8 +693,161 @@ contract MultiSubnetTest is Test, IntegrationTestBase {
+            _controllerSubnet: rootSubnetName
+        });
+
+        // the token controller sits in the root network.
+        ipcTokenController = new IpcTokenController({
+            _gateway: address(rootGateway),
+            _tokenContractAddress: address(testUSDC),
+            _destinationSubnet: nativeSubnetName,
+            _destinationContract: address(ipcTokenReplica)
+        });
+        ipcTokenReplica.setController(address(ipcTokenController));
+
+        vm.prank(holder);
+        testUSDC.approve(address(ipcTokenController), transferAmount);
+
+        console.log("mock usdc contract: %s", address(testUSDC));
+        console.log("mock usdc owner: %s", owner);
+        console.log("mock usdc holder: %s", address(holder));
+        console.log("ipcTokenController: %s", address(ipcTokenController));
+        console.log(
+            "controller allowance for holder: %d",
+            testUSDC.allowance(address(holder), address(ipcTokenController))
+        );
+
+        vm.prank(address(holder));
+        IpcEnvelope memory lockAndTransferEnvelope = ipcTokenController.lockAndTransferWithReturn(
+            recipient,
+            transferAmount
+        );
+
+        // Check that the message is in unconfirmedTransfers
+        (address receiptSender, uint256 receiptValue) = ipcTokenController.getUnconfirmedTransfer(
+            lockAndTransferEnvelope.toHash()
+        );
+        require(receiptSender == address(holder), "Transfer sender incorrect in unconfirmedTransfers");
+        require(receiptValue == transferAmount, "Transfer amount incorrect in unconfirmedTransfers");
+
+        //confirm that token replica only accept calls to Ipc from the gateway
+        vm.prank(owner);
+        vm.expectRevert(IpcHandler.CallerIsNotGateway.selector);
+        ipcTokenReplica.handleIpcMessage(expected);
+
+        // the message the root gateway's postbox is being executed in the token subnet's gateway
+
+        expected = IpcEnvelope({
+            kind: IpcMsgKind.Call,
+            from: IPCAddress({
+                subnetId: rootSubnetName,
+                rawAddress: FvmAddressHelper.from(address(ipcTokenController))
+            }),
+            to: lockAndTransferEnvelope.to,
+            value: 0,
+            message: lockAndTransferEnvelope.message,
+            nonce: 0 // nonce will be updated by LibGateway.commitCrossMessage
+        });
+
+        msgs[0] = expected;
+        executeTopDownMsgs(msgs, nativeSubnetName, address(nativeSubnetGateway));
+
+        //ensure that tokens are delivered on subnet
+        require(IERC20(ipcTokenReplica).balanceOf(recipient) == transferAmount, "incorrect proxy token balance");
+
+        console.log("--------------- withdraw token (bottom-up)---------------");
+
+        // ensure that USDC holder has 0 tokens in the root chain
+        // require(0 == testUSDC.balanceOf(holder), "unexpected holder balance in withdraw flow");
```

Comment:
A require like this makes sense to have, with the right balance (initial balance minus tokens previously sent).

Reply:
09a93ccb